File: hcs_toolbox_py/hproj.py
"""
Homogeneous projection onto the unit sphere.

Original MATLAB: hproj.m
"""

import numpy as np
from scipy.linalg import expm


def hproj(x, Gd, hn_fun):
    """
    Compute the homogeneous projection of a non-zero vector x onto the
    homogeneous unit sphere.

    Finds z and s such that:
        z = d(s) * x   and   hn_fun(z) = 1

    Parameters
    ----------
    x : (p,) or (p,1) ndarray
        Vector to be projected (must be non-zero).
    Gd : (p, p) ndarray
        Anti-Hurwitz matrix — generator of dilation d(s) = expm(s*Gd).
    hn_fun : callable
        Function that computes the homogeneous norm of a vector, e.g.,
        lambda v: hnorm(v, Gd, P).

    Returns
    -------
    z : (p,) ndarray
        Homogeneous projection of x (on the unit sphere).
    s : float
        Dilation parameter s = -ln(nx).
    """
    x = np.asarray(x).flatten()
    nx = hn_fun(x)
    if nx == 0.0:
        raise ValueError("Homogeneous projection: x must be non-zero.")
    s = -np.log(nx)
    z = expm(s * Gd) @ x
    return z, s

File: HCS_Toolbox_ver02/hproj.m
function [z, s]=hproj(x,Gd,hn_fun)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% The function [z, s]=hproj(x,Gd,hn_fun) computes homogeneous projection                   
%% of a given non-zero vector x on the homogeneous unit sphere, namely,
%% it finds a real vector z and real scalar s such that 
%%                                                            
%% z=d(s)x and the homogeneous norm of z equals 1  (x is non-zero)  
%%                                                            
%% The input parameters are                                   
%%     x  - vector (p x 1) which homogeneous projection has to be computed  
%%    Gd  - anti-Hurwitx matrix (p x p)  being the generator of
%%          the dilation d(s)=expm(s*Gd) for any real s 
%% hn_fun - a function which computes a homogenenous norm 
%% 
%%
%%
%% The function [z, s]=hproj(x, Gd, P, hn_fun, tol) uses the additional 
%% parameter tol to control compuational tolerance (by default, tol=1e-6)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

nx=hn_fun(x); s=0;z=0;
if nx==0 disp('Error: vector x must be nonzero'); 
else s=-log(nx); z=expm(s*Gd)*x; end;